packages/core/orchestration/src/transaction/distributed-transaction.ts

```diff
@@ -320,6 +316,76 @@ class DistributedTransaction extends EventEmitter {
   public hasTemporaryData(key: string) {
     return this.#temporaryStorage.has(key)
   }
+
+  /**
+   * Try to serialize the checkpoint data
+   * If it fails, it means that the context or the errors are not serializable
+   * and we should handle it
+   *
+   * @internal
+   * @returns
+   */
+  #serializeCheckpointData() {
+    const data = new TransactionCheckpoint(
+      this.getFlow(),
+      this.getContext(),
+      this.getErrors()
+    )
+
+    const isSerializable = (obj) => {
+      try {
+        JSON.parse(JSON.stringify(obj))
+        return true
+      } catch {
+        return false
+      }
+    }
+
+    let rawData
+    try {
+      rawData = JSON.parse(JSON.stringify(data))
+    } catch (e) {
+      if (!isSerializable(this.context)) {
+        // This is a safe guard, we should never reach this point
+        // If we do, it means that the context is not serializable
+        // and we should throw an error
+        throw new NonSerializableCheckPointError(
+          "Unable to serialize context object. Please make sure the workflow input and steps response are serializable."
+        )
+      }
+
+      if (!isSerializable(this.errors)) {
+        const nonSerializableErrors: TransactionStepError[] = []
+        for (const error of this.errors) {
+          if (!isSerializable(error.error)) {
+            error.error = {
+              name: error.error.name,
+              message: error.error.message,
+              stack: error.error.stack,
+            }
+            nonSerializableErrors.push({
+              ...error,
+              error: e,
+            })
+          }
+        }
+
+        if (nonSerializableErrors.length) {
+          this.errors.push(...nonSerializableErrors)
```

Comment:
shouldn't the existing non-serialisable errors be removed?

Reply:
We keep it but we cherry pick useful info so that you still have access to it if needed